File: Day05/main.py

# quick and dirty

type Rule = tuple[int, int]
type Rules = set[Rule]

type Pages = list[int]

def read_input() -> tuple[Rules, list[Pages]]:
    with open("input.txt") as file:
        lines = [ line.rstrip('\n') for line in file.readlines() ]
        i = lines.index('')
        rules = lines[:i]
        def convert_rule(rule: str) -> Rule:
            [x, y] = rule.split('|')
            return (int(x), int(y))
        rules = { convert_rule(rule) for rule in rules }
        pages_list = lines[i+1:]
        pages_list = [ list(map(int, pages.split(','))) for pages in pages_list ]
    return (rules, pages_list)

def find_violating_rules(rules, pages):
    n = len(pages)
    return ( rule for i in range(n) for j in range(i + 1, n) if (rule := (pages[j], pages[i])) in rules )

def is_correctly_ordered(rules, pages) -> bool:
    return not any(find_violating_rules(rules, pages))

def part1():
    (rules, pages_list) = read_input()
    answer = 0
    for pages in pages_list:
        if is_correctly_ordered(rules, pages):
            answer += pages[len(pages)//2]
    print(answer)

def part2():
    (rules, pages_list) = read_input()
    answer = 0
    for pages in pages_list:
        if is_correctly_ordered(rules, pages):
            continue
        n = len(pages)
        while not is_correctly_ordered(rules, pages):
            rule = next(find_violating_rules(rules, pages))
            i, j = pages.index(rule[0]), pages.index(rule[1])
            pages[i], pages[j] = pages[j], pages[i]
        answer += pages[len(pages)//2]
    print(answer)

if __name__ == '__main__':
    part1()
    part2()

File: Day05/main.hs
import Data.Set (Set)
import qualified Data.Set as Set


-- list helpers

split :: (a -> Bool) -> [a] -> [[a]]
split p xs = case break p xs of
  (xs, [])   -> [xs]
  (xs, _:ys) -> xs : split p ys

middle :: [a] -> a
middle [x, y, z] = y
middle ls = middle (tail . init $ ls)

swap :: Eq a => (a, a) -> [a] -> [a]
swap _ [] = []
swap (a, b) (x:xs) = x' : swap (a, b) xs where
  x'
    | x == a    = b
    | x == b    = a
    | otherwise = x


-- parsing

type Rule = (Int, Int)

type Update = [Int]

parseInput :: IO (Set Rule, [Update])
parseInput = do
  contents <- readFile "input.txt"
  let [rules, updates] = split null (lines contents)
  let parseRule :: String -> Rule
      parseRule = (\[a, b] -> (read a, read b)) . split (== '|')
  let parseUpdate :: String -> Update
      parseUpdate = map read . split (== ',')
  return (Set.fromList (map parseRule rules), map parseUpdate updates)


-- helpers

impliedRules :: Update -> [Rule]
impliedRules [] = []
impliedRules (p:ps) = [ (p, p') | p' <- ps ] ++ impliedRules ps

flipRule :: Rule -> Rule
flipRule (a, b) = (b, a)

violatedRules :: Set Rule -> Update -> [Rule]
violatedRules rules update = [ violatedRule | rule <- impliedRules update, let violatedRule = flipRule rule, violatedRule `Set.member` rules ]

isCorrectlyOrdered :: Set Rule -> Update -> Bool
isCorrectlyOrdered rules update = null (violatedRules rules update)

part1 :: IO ()
part1 = do
  (rules, updates) <- parseInput
  let answer = sum [ middle update | update <- updates, isCorrectlyOrdered rules update ]
  print answer

part2 :: IO ()
part2 = do
  (rules, updates) <- parseInput
  let incorrectlyOrderedUpdates = [ update | update <- updates, not (isCorrectlyOrdered rules update) ]
  let fixOrder :: Update -> Update
      fixOrder update = case violatedRules rules update of
        []       -> update
        (rule:_) -> fixOrder (swap rule update)
  let answer = sum [ middle (fixOrder update) | update <- incorrectlyOrderedUpdates ]
  print answer

main :: IO ()
main = do
  part1
  part2
